openvdb/openvdb/points/PointReplicate.h

```diff
@@ -0,0 +1,337 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+
+/// @author Nick Avramoussis
+///
+/// @file PointReplicate.h
+///
+/// @brief  Algorithms to duplicate points in PointDataGrids.
+
+#ifndef OPENVDB_POINTS_POINT_REPLICATE_HAS_BEEN_INCLUDED
+#define OPENVDB_POINTS_POINT_REPLICATE_HAS_BEEN_INCLUDED
+
+#include <openvdb/points/PointDataGrid.h>
+#include <openvdb/tools/Prune.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+
+/// @brief Replicates points provided in a source grid into a new grid,
+///        transfering and creating attributes found in a provided
+///        attribute vector. If an attribute doesn't exist, it is ignored.
+///        Position is always replicated, leaving the new points exactly
+///        over the top of the source points.
+/// @note  The position attribute must exist
+/// @param source      The source grid to replicate points from
+/// @param multiplier  The base number of points to replicate per point
+/// @param attributes  Attributes to transfer to the new grid
+/// @param scaleAttribute  A scale float attribute which multiplies the base
+///                        multiplier to vary the point count per point.
+/// @param replicationIndex  When provided, creates a replication attribute
```

Comment:
Could we rename `replicationIndex` to `targetIndex` and it would be good to add an optional `sourceIndex` as well, that records the index of the source point the new target point was replicated from?

I also think it might be more intuitive to either have the global `multiplier` version of this method or to provide an integer `countAttribute` instead of a float `scaleAttribute`. It's relatively little work for the user to compute that directly and avoids the ambiguity of what happens if multipler * scale is not a whole number. The point scatter also has a float value, but interprets it quite differently by generating a random value to determine if it should generate the rounded down or rounded up value - .https://github.com/AcademySoftwareFoundation/openvdb/blob/master/openvdb/openvdb/points/PointScatter.h#L85.

Reply:
> Could we rename replicationIndex to targetIndex and it would be good to add an optional sourceIndex as well, that records the index of the source point the new target point was replicated from?

You can achieve this yourself by transfering an `id` attribute from the source points!

> I also think it might be more intuitive to either have the global multiplier version of this method or to provide an integer countAttribute instead of a float scaleAttribute. 

Pretty sure I chose the current way as this is what Houdini does